File: python/14_fbtree.py

"""
Check if a binary tree is a full binary tree in Python.
"""

from typing import Optional

class Node:
    """
    A node in a binary tree.
    """
    def __init__(self, item: int) -> None:
        """Initialize a tree node with value and left/right children."""
        self.item: int = item
        self.leftChild: Optional['Node'] = None
        self.rightChild: Optional['Node'] = None

def is_full_tree(root: Optional[Node]) -> bool:
    """
    Check if the binary tree rooted at 'root' is a full binary tree.
    """
    if root is None:
        return True
    if root.leftChild is None and root.rightChild is None:
        return True
    if root.leftChild is not None and root.rightChild is not None:
        return is_full_tree(root.leftChild) and is_full_tree(root.rightChild)
    return False


if __name__ == "__main__":
    root = Node(1)
    root.rightChild = Node(3)
    root.leftChild = Node(2)
    root.leftChild.leftChild = Node(4)
    root.leftChild.rightChild = Node(5)
    root.leftChild.rightChild.leftChild = Node(6)
    root.leftChild.rightChild.rightChild = Node(7)
    if is_full_tree(root):
        print("The tree is a full binary tree")
    else:
        print("The tree is not a full binary tree")

File: java/14_fbtree.java
/**
 * Check whether a binary tree is a full binary tree. A full binary tree is a
 * tree in which every node has either 0 or 2 children.
 */
class Node {
    int data;
    Node leftChild, rightChild;

    Node(int item) {
        data = item;
        leftChild = rightChild = null;
    }
}

class BinaryTree {
    Node root;

    /**
     * Determine whether the subtree rooted at {@code node} is a full binary
     * tree. The method is recursive and preserves the original algorithm.
     *
     * @param node subtree root (may be null)
     * @return true if the subtree is full, or if {@code node} is null
     */
    boolean isFullBinaryTree(Node node) {
        // An empty tree is considered full
        if (node == null) {
            return true;
        }

        // A leaf node is full
        if (node.leftChild == null && node.rightChild == null) {
            return true;
        }

        // If both left and right are non-null, check recursively
        if (node.leftChild != null && node.rightChild != null) {
            return isFullBinaryTree(node.leftChild) && isFullBinaryTree(node.rightChild);
        }

        // One child only -> not full
        return false;
    }

    /**
     * Convenience wrapper: check whether the whole tree (root) is full.
     */
    public boolean isFull() {
        return isFullBinaryTree(root);
    }

    public static void main(String args[]) {
        BinaryTree tree = new BinaryTree();
        tree.root = new Node(1);
        tree.root.leftChild = new Node(2);
        tree.root.rightChild = new Node(3);
        tree.root.leftChild.leftChild = new Node(4);
        tree.root.leftChild.rightChild = new Node(5);
        tree.root.rightChild.leftChild = new Node(6);
        tree.root.rightChild.rightChild = new Node(7);

        if (tree.isFull()) {
            System.out.print("The tree is a full binary tree");
        } else {
            System.out.print("The tree is not a full binary tree");
        }
    }
}